[depends] qt: "Project ERROR: failed to parse default search paths from compiler output" on macOS 14.0

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

Running `make -C depends qt` on macOS 14.0 Sonoma produces the following error:

```
Extracting qt...
/Users/.../bitcoin/depends/sources/qtbase-everywhere-opensource-src-5.15.5.tar.xz: OK
/Users/.../bitcoin/depends/sources/qttranslations-everywhere-opensource-src-5.15.5.tar.xz: OK
/Users/.../bitcoin/depends/sources/qttools-everywhere-opensource-src-5.15.5.tar.xz: OK
Preprocessing qt...
patching file 'qtbase/configure'
patching file 'qtbase/mkspecs/features/qt_module.prf'
patching file 'qtbase/src/plugins/platforms/android/androidjnimain.cpp'
patching file 'qtbase/src/plugins/platforms/xcb/qxcbcursor.cpp'
patching file 'qtbase/configure.json'
patching file 'qtbase/configure.pri'
patching file 'qtbase/mkspecs/common/macx.conf'
patching file 'qtbase/mkspecs/features/mac/default_post.prf'
patching file 'qtbase/mkspecs/features/toolchain.prf'
patching file 'qtbase/src/tools/moc/main.cpp'
patching file 'qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h'
patching file 'qtbase/mkspecs/features/android/default_pre.prf'
patching file 'qtbase/src/tools/rcc/rcc.cpp'
patching file 'qtbase/src/gui/text/qplatformfontdatabase.cpp'
patching file 'qtbase/src/gui/text/qplatformfontdatabase.h'
patching file 'qtbase/src/platformsupport/fontdatabases/mac/qfontengine_coretext.mm'
patching file 'qtbase/src/platformsupport/fontdatabases/mac/qfontengine_coretext_p.h'
patching file 'qtbase/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp'
patching file 'qtbase/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase_p.h'
patching file 'qtbase/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp'
patching file 'qtbase/src/3rdparty/double-conversion/fixed-dtoa.h'
patching file 'qtbase/mkspecs/features/toolchain.prf'
Configuring qt...
Creating qmake...
Info: creating super cache file /Users/.../bitcoin/depends/work/build/x86_64-apple-darwin23.0.0/qt/5.15.5-723c1c55951/.qmake.super
Info: creating cache file /Users/.../bitcoin/depends/work/build/x86_64-apple-darwin23.0.0/qt/5.15.5-723c1c55951/.qmake.cache
Info: creating stash file /Users/.../bitcoin/depends/work/build/x86_64-apple-darwin23.0.0/qt/5.15.5-723c1c55951/.qmake.stash
Command line: -bindir /Users/.../bitcoin/depends/x86_64-apple-darwin23.0.0/native/bin -c++std c++17 -confirm-license -hostprefix /Users/.../bitcoin/depends/x86_64-apple-darwin23.0.0/native -no-compile-examples -no-cups -no-egl -no-eglfs -no-evdev -no-gif -no-glib -no-icu -no-ico -no-iconv -no-kms -no-linuxfb -no-libjpeg -no-libproxy -no-libudev -no-mimetype-database -no-mtdev -no-openssl -no-openvg -no-reduce-relocations -no-schannel -no-sctp -no-securetransport -no-sql-db2 -no-sql-ibase -no-sql-oci -no-sql-tds -no-sql-mysql -no-sql-odbc -no-sql-psql -no-sql-sqlite -no-sql-sqlite2 -no-system-proxies -no-use-gold-linker -no-zstd -nomake examples -nomake tests -nomake tools -opensource -pkg-config -prefix /Users/.../bitcoin/depends/x86_64-apple-darwin23.0.0 -qt-libpng -qt-pcre -qt-harfbuzz -qt-zlib -static -v -no-feature-bearermanagement -no-feature-colordialog -no-feature-commandlineparser -no-feature-concurrent -no-feature-dial -no-feature-fontcombobox -no-feature-ftp -no-feature-http -no-feature-image_heuristic_mask -no-feature-keysequenceedit -no-feature-lcdnumber -no-feature-networkdiskcache -no-feature-networkproxy -no-feature-pdf -no-feature-printdialog -no-feature-printer -no-feature-printpreviewdialog -no-feature-printpreviewwidget -no-feature-sessionmanager -no-feature-socks5 -no-feature-sql -no-feature-sqlmodel -no-feature-statemachine -no-feature-syntaxhighlighter -no-feature-textbrowser -no-feature-textmarkdownwriter -no-feature-textodfwriter -no-feature-topleveldomain -no-feature-udpsocket -no-feature-undocommand -no-feature-undogroup -no-feature-undostack -no-feature-undoview -no-feature-vnc -no-feature-wizard -no-feature-xml -release -silent -no-dbus -no-opengl -pch -no-feature-corewlan -no-freetype QMAKE_MACOSX_DEPLOYMENT_TARGET=11.0
Project ERROR: failed to parse default search paths from compiler output
make: *** [/Users/.../bitcoin/depends/x86_64-apple-darwin23.0.0/.qt_stamp_configured] Error 3
```

### Expected behaviour

Dependency compilation should work out of the box.

### Steps to reproduce

Check out a fresh repo on a macOS 14.0 system and run `make -C depends qt`

### Relevant log output

_No response_

### How did you obtain Bitcoin Core

Compiled from source

### What version of Bitcoin Core are you using?

git latest

### Operating system and version

macOS 14.0 Sonoma

### Machine specifications

Both arm64 and x86_64 macs affected.

--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -22,6 +22,7 @@ $(package)_patches += rcc_hardcode_timestamp.patch
 $(package)_patches += duplicate_lcqpafonts.patch
 $(package)_patches += fast_fixed_dtoa_no_optimize.patch
 $(package)_patches += guix_cross_lib_path.patch
+$(package)_patches += fix-macos-linker.patch
 
 $(package)_qttranslations_file_name=qttranslations-$($(package)_suffix)
 $(package)_qttranslations_sha256_hash=c92af4171397a0ed272330b4fa0669790fcac8d050b07c8b8cc565ebeba6735e
@@ -238,6 +239,7 @@ endef
 define $(package)_preprocess_cmds
   cp $($(package)_patch_dir)/qt.pro qt.pro && \
   cp $($(package)_patch_dir)/qttools_src.pro qttools/src/src.pro && \
+  patch -p1 -i $($(package)_patch_dir)/fix-macos-linker.patch && \
   patch -p1 -i $($(package)_patch_dir)/dont_hardcode_pwd.patch && \
   patch -p1 -i $($(package)_patch_dir)/fix_qt_pkgconfig.patch && \
   patch -p1 -i $($(package)_patch_dir)/fix_android_jni_static.patch && \

--- a/depends/patches/qt/fix-macos-linker.patch
+++ b/depends/patches/qt/fix-macos-linker.patch
@@ -1,0 +1,55 @@
+qmake: Don't error out if QMAKE_DEFAULT_LIBDIRS is empty on macOS
+
+The new linker in Xcode 15 doesn't provide any default linker or
+framework paths when requested via -v, but still seems to use the
+default paths documented in the ld man page.
+
+We trust that linker will do the right thing, even if we don't
+know of its default linker paths.
+
+We also need to opt out of the default fallback logic to
+set the libdirs to /lib and /usr/lib.
+
+This may result in UnixMakefileGenerator::findLibraries finding
+different libraries than expected, if additional paths are
+passed with -L, which will then take precedence for qmake,
+even if the linker itself will use the library from the
+SDK's default paths. This should hopefully not be an issue
+in practice, as we don't turn -lFoo into absolute paths in
+qmake, so the only risk is that we're picking up the wrong
+prl files and adding additional dependencies that the lib
+in the SDK doesn't have.
+
+Upstream commits:
+ - Qt 5.15.16: Not yet publicly available.
+ - Qt dev: cdf64b0e47115cc473e1afd1472b4b09e130b2a5
+
+For other Qt branches see
+https://codereview.qt-project.org/q/I2347b26e2df0828471373b0e15b8c9089274c65d
+
+--- old/qtbase/mkspecs/features/toolchain.prf
++++ new/qtbase/mkspecs/features/toolchain.prf
+@@ -283,9 +283,12 @@ isEmpty($${target_prefix}.INCDIRS) {
+                 }
+             }
+         }
+-        isEmpty(QMAKE_DEFAULT_LIBDIRS)|isEmpty(QMAKE_DEFAULT_INCDIRS): \
++        isEmpty(QMAKE_DEFAULT_INCDIRS): \
+             !integrity: \
+-                error("failed to parse default search paths from compiler output")
++                error("failed to parse default include paths from compiler output")
++        isEmpty(QMAKE_DEFAULT_LIBDIRS): \
++            !integrity:!darwin: \
++                error("failed to parse default library paths from compiler output")
+         QMAKE_DEFAULT_LIBDIRS = $$unique(QMAKE_DEFAULT_LIBDIRS)
+     } else: ghs {
+         cmd = $$QMAKE_CXX $$QMAKE_CXXFLAGS -$${LITERAL_HASH} -o /tmp/fake_output /tmp/fake_input.cpp
+@@ -407,7 +410,7 @@ isEmpty($${target_prefix}.INCDIRS) {
+         QMAKE_DEFAULT_INCDIRS = $$split(INCLUDE, $$QMAKE_DIRLIST_SEP)
+     }
+ 
+-    unix:if(!cross_compile|host_build) {
++    unix:!darwin:if(!cross_compile|host_build) {
+         isEmpty(QMAKE_DEFAULT_INCDIRS): QMAKE_DEFAULT_INCDIRS = /usr/include /usr/local/include
+         isEmpty(QMAKE_DEFAULT_LIBDIRS): QMAKE_DEFAULT_LIBDIRS = /lib /usr/lib
+     }

